### PR TITLE
 gateway-api: Support GRPCRoute resource

### LINKED ISSUE #21928: CFP: Support GRPCRoute in Cilium Gateway API
## Cilium Feature Proposal
        
The first cut for Gateway API was done in https://github.com/cilium/cilium/pull/21749, however, we have a couple of follow-up tasks to support more features from Gateway API upstream.

This is to track the work to support [GRPCRoute](https://gateway-api.sigs.k8s.io/api-types/grpcroute/).

## Goals

Based on the progress and code re-usability introduced in [TLSRoutes](https://github.com/cilium/cilium/issues/21927) we should build a similar implementation of GRPCRoute. This route should re-use logic from both TLS and HTTP routes to route gRPC. 

To implement it we should re-use the Envoy logic as an GRPCRoute can be attached to a HTTP(S) Gateway, we can therefore re-use that for TLS termination. Preferably it should use gRPC specific routing in Envoy where possible. 

## Proposed Solution

TBD

## Tasks

- [ ] Add required reconcile functions to validate GRPCRoute
  - [ ] Validate against hostname re-use between HTTP and gRPC
- [ ] Add a translator module for GRPCRoutes rules to Envoy rules
  - [ ] Consider a gRPC Web filter https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/grpc_web_filter & https://github.com/grpc/grpc-web/blob/master/net/grpc/gateway/examples/echo/envoy.yaml
- [ ] Handle edge cases according to spec (TBD)
- [ ] Validate required CRDs will be installed
- [ ] Pass conformance tests
- [ ] Awesome documentation!

### PR DESCRIPTION
**Pull request by sayboras**
### Description

Please refer to individual commits for more details.

Fixes: https://github.com/cilium/cilium/issues/21928

```release-note
gateway-api: Support GRPCRoute resource
```

### CODE REVIEW

**Review by qmonnet — APPROVED** [review]
I came for the docs, and the docs looked good.

**Review by nbusseneau — APPROVED** [review]
Trivial ack for ci-structure.

**Review by squeed — APPROVED** [review]

File: operator/pkg/gateway-api/grpcroute.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+					)
+				}
+			}
+			return backendServices
+		},
+	); err != nil {
+		return err
+	}
+
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &gatewayv1alpha2.GRPCRoute{}, gatewayIndex,
```
Comment (squeed) [inline operator/pkg/gateway-api/grpcroute.go]: This line is is some proper arcane incantation. Can you please add a comment :-)?

File: operator/pkg/gateway-api/grpcroute_reconcile.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	gatewayv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/cilium/cilium/operator/pkg/gateway-api/routechecks"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
```
Comment (squeed) [inline operator/pkg/gateway-api/grpcroute_reconcile.go:23]: Please add a comment giving a basic overview, especially the inputs and outputs.

File: .github/workflows/conformance-gateway-api.yaml [hunk 3]
```diff
@@ -169,6 +169,7 @@ jobs:
           kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/${{ env.gateway_api_version }}/config/crd/${{ matrix.crd-channel }}/gateway.networking.k8s.io_referencegrants.yaml
           ## TLSRoute is only available in experimental channel in v0.7.0
           kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/${{ env.gateway_api_version }}/config/crd/experimental/gateway.networking.k8s.io_tlsroutes.yaml
+          kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/${{ env.gateway_api_version }}/config/crd/experimental/gateway.networking.k8s.io_grpcroutes.yaml
 
           # To make sure that Gateway API CRs are available
```
Comment (mhofstetter) [inline .github/workflows/conformance-gateway-api.yaml:174]: shouldn't we add a check for the `GrpcRoute`?

File: operator/pkg/gateway-api/grpcroute.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+		func(rawObj client.Object) []string {
+			hr := rawObj.(*gatewayv1alpha2.GRPCRoute)
```
Comment (mhofstetter) [inline operator/pkg/gateway-api/grpcroute.go]: nit: `hr` -> `gr` (or just `route`)

File: operator/pkg/gateway-api/grpcroute.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+	client.Client
+	Scheme *runtime.Scheme
+
+	Model *internalModel
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *grpcRouteReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &gatewayv1alpha2.GRPCRoute{}, backendServiceIndex,
+		func(rawObj client.Object) []string {
```
Comment (mhofstetter) [inline operator/pkg/gateway-api/grpcroute.go]: might be worth to define the index functions as normal functions (and reference them here) instead of anonymous functions. should be easier to read and document them.

**Review by joamaki — APPROVED** [review]
Approve for `install/kubernetes` changes.

File: operator/pkg/gateway-api/grpcroute.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+			hr, ok := rawObj.(*gatewayv1alpha2.GRPCRoute)
```
Comment (mhofstetter) [inline operator/pkg/gateway-api/grpcroute.go]: nit: `hr` -> `gr` (or just `route`)

File: operator/pkg/gateway-api/gateway_reconcile.go [hunk 7]
```diff
@@ -226,6 +233,16 @@ func (r *gatewayReconciler) filterHTTPRoutesByGateway(ctx context.Context, gw *g
 	return filtered
 }
 
+func (r *gatewayReconciler) filterGRPCRoutesByGateway(ctx context.Context, gw *gatewayv1beta1.Gateway, routes []gatewayv1alpha2.GRPCRoute) []gatewayv1alpha2.GRPCRoute {
```
Comment (mhofstetter) [inline operator/pkg/gateway-api/gateway_reconcile.go]: maybe out of scope for this PR: we should start thinking about a way to get rid of this boilerplate for every route-type.

File: .github/workflows/conformance-gateway-api.yaml [hunk 8]
```diff
@@ -169,6 +169,7 @@ jobs:
           kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/${{ env.gateway_api_version }}/config/crd/${{ matrix.crd-channel }}/gateway.networking.k8s.io_referencegrants.yaml
           ## TLSRoute is only available in experimental channel in v0.7.0
           kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/${{ env.gateway_api_version }}/config/crd/experimental/gateway.networking.k8s.io_tlsroutes.yaml
+          kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/gateway-api/${{ env.gateway_api_version }}/config/crd/experimental/gateway.networking.k8s.io_grpcroutes.yaml
```
Comment (mhofstetter) [inline .github/workflows/conformance-gateway-api.yaml:172]: do we have to add the grpcRoute to the list of supported features to include this in the gateway-api conformance tests?

https://github.com/cilium/cilium/blob/7add21c7e44386cb5d9f72954bf3b911e4ce242a/.github/workflows/conformance-gateway-api.yaml#L121

edit: or is this the gRPC Route not yet covered by the conformance tests or included in the http route? :thinking: https://gateway-api.sigs.k8s.io/geps/gep-917/#conformance-profiles

**Review by mhofstetter — COMMENTED** [review]
nice work @sayboras :partying_face: 

i took a first look and added some comments inline. I'm not sure whether the gateway-api conformance tests are already enabled without having added the grpcRoute feature to the list of supported features. Might be worth to activate that first and have some feedback from the conformance tests too :smile: (comment: https://github.com/cilium/cilium/pull/28654#discussion_r1365137956)

File: .github/workflows/conformance-gateway-api.yaml (hunk 3 above)
Comment (sayboras) [reply .github/workflows/conformance-gateway-api.yaml:174]: ah good catch :100:

File: operator/pkg/gateway-api/gateway_reconcile.go (hunk 7 above)
Comment (sayboras) [reply operator/pkg/gateway-api/gateway_reconcile.go]: yeah, we might need to wrap the types, similar to what we have with input check. 

Will do this as a follow up.

File: .github/workflows/conformance-gateway-api.yaml (hunk 8 above)
Comment (sayboras) [reply .github/workflows/conformance-gateway-api.yaml:172]: yup, there is no conformance test right now, however, the translation logic is well covered in HTTP conformance tests, which gives us high enough confidence :pray:

Comment (mhofstetter) [reply .github/workflows/conformance-gateway-api.yaml:172]: But i assume the HTTP conformance tests don't make use of the CRD `GRPCRoute`? if not we should at least introduce such a test (e.g. in the connectivity tests) as a follow up. WDYT?

Comment (sayboras) [reply .github/workflows/conformance-gateway-api.yaml:172]: yup, makes sense to add some basic grpcroute test instead of doing it manually :+1:

Comment (sayboras) [reply .github/workflows/conformance-gateway-api.yaml:172]: let me do it in this PR as well, marking this draft to add grpc test

Comment (sayboras) [reply .github/workflows/conformance-gateway-api.yaml:172]: done, one simple grpc echo route, and curl with grpc web steps are added.

Comment (mhofstetter) [reply .github/workflows/conformance-gateway-api.yaml:172]: nice, thanks!

**Review by mhofstetter — APPROVED** [review]
LGTM, thanks again!
